### ISSUE TITLE
build-cli-bin fails with invalid pseudo-version: does not match version-control timestamp

### ISSUE THREAD

**Issue by jeffreyfriedman** [linkerd/linkerd2#3566]
## Bug Report
After a new git clone, attempting to execute `bin/build-cli-bin` fails.

### What is the issue?
A dependency appears to have an invalid timestamp compared to the tagged version.

### How can it be reproduced?

Clone the repo, then execute `bin/build-cli-bin`.  May require Go 1.13.1 to reproduce.

### Logs, error output, etc

```
➜  linkerd2 git:(master) bin/build-cli-bin
go: github.com/deislabs/smi-sdk-go@v0.0.0-20190610232231-f281e2121a16 requires
	k8s.io/apimachinery@v0.0.0-20181127105237-2b1284ed4c93: invalid pseudo-version: does not match version-control timestamp (2018-11-27T02:52:37Z)
```

#### `linkerd check` output

```
➜  linkerd2 git:(master) linkerd check
INFO: 2019/10/13 08:48:09 CPU time info is unavailable on non-linux or appengine environment.
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ no invalid service profiles

linkerd-version
---------------
√ can determine the latest version
‼ cli is up-to-date
    is running version 2.5.0 but the latest stable version is 2.6.0
    see https://linkerd.io/checks/#l5d-version-cli for hints

control-plane-version
---------------------
√ control plane is up-to-date
‼ control plane and cli versions match
    control plane running stable-2.6.0 but cli running stable-2.5.0
    see https://linkerd.io/checks/#l5d-version-control for hints

Status check results are √
```

### Environment

- Kubernetes Version: N/A (local OSX machine)
- Cluster Environment: N/A (local OSX machine)
- Host OS: OSX
- Linkerd version: attempting to run 2.6

### Additional context
This might be a Go 1.13 issue: https://tip.golang.org/doc/go1.13#version-validation.  

It's breaking the formula version bump for Homebrew, per https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/49724/version=catalina/testReport/brew-test-bot/catalina/install___build_bottle_linkerd/

```
➜  linkerd2 git:(master) go version
go version go1.13.1 darwin/amd64
```

**Comment by srv-twry** [comment]
Hi @jeffreyfriedman, I recently faced the same issue as well. I was using `go 1.13`. I can confirm that the issue was the use of version `1.13`. Installing version `1.12.9` resolved it.

Your additional context is spot on. Go `1.13` does version validations on transitive dependencies as well.

Basically, the project's `go.mod` file declares a dependency on `k8s.io/apimachinery v0.0.0-20190612205821-1799e75a0719`. The file also declares a dependency on the `github.com/deislabs/smi-sdk-go v0.0.0-20190610232231-f281e2121a16` which has it's own dependency on `k8s.io/apimachinery v0.0.0-20181127025237-2b1284ed4c93`. So as a result, there are two dependencies(one direct and other transitive) on the **apimachinery** package with different versions.

**Comment by alpeb** [comment]
It's normal for the project to declare a dependency that is also required transitively but using a different version (it would be too much of a coincidence that both versions would match). From what I understand, go just resolves to use the most recent one.

*TL;DR;*
To support go 1.13, `smi-sdk` and `go-client` need to be updated, but their updates clash with one another.

-----

I did try to fix the invalid pseudo version as the [doc explains](https://tip.golang.org/doc/go1.13#version-validation), but for some reason the same invalid version keeps on creeping in.

So instead of that I tried to upgrade the `smi-sdk-go` dependency to exactly the [sha](https://github.com/deislabs/smi-sdk-go/commit/9ef7916f8ab24a47f3862241aed24880613662a8) where they fixed that versioning issue.

The next problem was:
```
github.com/Azure/go-autorest/autorest/adal: ambiguous import: found
github.com/Azure/go-autorest/autorest/adal in multiple modules:
github.com/Azure/go-autorest v11.3.2+incompatible (/home/alpeb/go/pkg/mod/github.com/!azure/go-autorest@v11.3.2+incompatible/autorest/adal)
github.com/Azure/go-autorest/autorest/adal v0.1.0 (/home/alpeb/go/pkg/mod/github.com/!azure/go-autorest/autorest/adal@v0.1.0)
```
At some point the `adal` package moved from `github.com/Azure/go-autorest/` to `github.com/Azure/go-autorest/autorest`.
`go-client` still depended on the former, while some other dependencies moved on, but go's dependency system doesn't grok that.
`go-client` solved that in this [sha](https://github.com/kubernetes/client-go/commit/fc8f647b3e527211f9f1f5491907fa814c303068) which shipped with `go-client@kubernetes-1.16.0`.

But after updating `go-client` we get this error:
```
github.com/deislabs/smi-sdk-go@v0.0.0-20190816171600-9ef7916f8ab2/pkg/gen/c=
lient/split/clientset/versioned/typed/split/v1alpha1/split_client.go:72:32:
undefined: serializer.DirectCodecFactory
```
The cause being the new `go-client` pulls `apimachinery@kubernetes-1.16.0` as well, which clashes with `smi-sdk` because it [requires](https://github.com/deislabs/smi-sdk-go/blob/master/pkg/gen/client/split/clientset/versioned/typed/split/v1alpha2/split_client.go#L72) an apimachinery struct that[ no longer exists in `1.16.0`](https://github.com/kubernetes/apimachinery/commit/4fac835274d8f4a7de6666ac9544f793d2884b84#diff-71b26e1e133ec6d3c4da26366b6502ac).

So our support for go 1.13 is blocked by `smi-sdk` updating their `apimachinery` dependency, and regenerating their [k8s client](https://github.com/deislabs/smi-sdk-go/tree/master/pkg/gen/client).
When they do, we'll also have to regenerate [our own](https://github.com/linkerd/linkerd2/tree/master/controller/gen/client), and move our smi `trafficsplit` dependency to `v1alpha2`.
Also, our `stern` fork will require a `go-client` upgrade as well.

**Comment by siggy** [comment]
@alpeb In case you have not already tried this, I got another project working by adding this to the bottom of `go.mod`:
```
replace k8s.io/apimachinery v0.0.0-20181127105237-2b1284ed4c93 => k8s.io/apimachinery v0.0.0-20181127025237-2b1284ed4c93

replace k8s.io/code-generator v0.0.0-20181116203124-c2090bec4d9b => k8s.io/code-generator v0.0.0-20181117043124-c2090bec4d9b
```

**Comment by alpeb** [comment]
Thanks @siggy that did the trick!
What I had tried was (following the docs, omitting the fixed date for go to fetch the appropriate one):
```
replace k8s.io/apimachinery v0.0.0-20181127105237-2b1284ed4c93 => k8s.io/apimachinery 2b1284ed4c93
```
but apparently that mechanism is broken.

After that, I had to fix a unit test, and I'm going through the integration tests to see if we can finally upgrade to go 1.13.

**Comment by siggy** [comment]
@alpeb Interesting! The omitting-the-fixed-date technique you mentioned is exactly how I arrived at those replace directives. I was using Go 1.13.4, is it possible you were on an older 1.13.x version?

**Comment by alpeb** [comment]
That's weird, I'm also in 1.13.4 :man_shrugging:  :
```bash
$ go version
go version go1.13.4 linux/amd64
$ go build
go: github.com/deislabs/smi-sdk-go@v0.0.0-20190610232231-f281e2121a16 requires
	k8s.io/apimachinery@v0.0.0-20181127105237-2b1284ed4c93: invalid pseudo-version: does not match version-control timestamp (2018-11-27T02:52:37Z)
$ echo "replace k8s.io/apimachinery@v0.0.0-20181127105237-2b1284ed4c93 => k8s.io/apimachinery 2b1284ed4c93" >> go.mod
$ go build
go: finding k8s.io/apimachinery 2b1284ed4c93
go: github.com/deislabs/smi-sdk-go@v0.0.0-20190610232231-f281e2121a16 requires
	k8s.io/apimachinery@v0.0.0-20181127105237-2b1284ed4c93: invalid pseudo-version: does not match version-control timestamp (2018-11-27T02:52:37Z)
```
